src/thor/dijkstras.cc

```diff
@@ -154,34 +159,71 @@ void Dijkstras::ExpandForward(GraphReader& graphreader,
     // (based on the costing method) or if a complex restriction exists for
     // this path.
     if (directededge->is_shortcut() || es->set() == EdgeSet::kPermanent ||
-        !(directededge->forwardaccess() & access_mode_)) {
+        !((FORWARD ? directededge->forwardaccess() : directededge->reverseaccess()) & access_mode_)) {
       continue;
     }
 
+    // Get end node tile, opposing edge Id, and opposing directed edge.
+    graph_tile_ptr t2;
+    baldr::GraphId oppedgeid;
+    const baldr::DirectedEdge* opp_edge;
+    // Only populate these at this point for the reverse search - the forward search
+    // may exit before needing them, so we can defer some work until later
+    if (!FORWARD) { // aka reverse
+      t2 = tile;
+      oppedgeid = graphreader.GetOpposingEdgeId(edgeid, t2);
+      if (t2 == nullptr) {
+        continue;
+      }
+      opp_edge = t2->directededge(oppedgeid);
+    }
+
     // Check if the edge is allowed or if a restriction occurs
     EdgeStatus* todo = nullptr;
     uint8_t restriction_idx = -1;
     if (offset_time.valid) {
       // With date time we check time dependent restrictions and access
-      if (!costing_->Allowed(directededge, pred, tile, edgeid, offset_time.local_time,
-                             nodeinfo->timezone(), restriction_idx) ||
-          costing_->Restricted(directededge, pred, bdedgelabels_, tile, edgeid, true, todo,
-                               offset_time.local_time, nodeinfo->timezone())) {
+      const bool allowed =
+          FORWARD ? costing_->Allowed(directededge, pred, tile, edgeid, offset_time.local_time,
+                                      nodeinfo->timezone(), restriction_idx)
+                  : costing_->AllowedReverse(directededge, pred, opp_edge, t2, oppedgeid,
+                                             offset_time.local_time, nodeinfo->timezone(),
+                                             restriction_idx);
+      if (!allowed || costing_->Restricted(directededge, pred, bdedgelabels_, tile, edgeid, true,
+                                           todo, offset_time.local_time, nodeinfo->timezone())) {
         continue;
       }
     } else {
-      if (!costing_->Allowed(directededge, pred, tile, edgeid, 0, 0, restriction_idx) ||
-          costing_->Restricted(directededge, pred, bdedgelabels_, tile, edgeid, true)) {
+      const bool allowed =
```

Comment:
Should we combine the `if(offset_time.valid) ...else` block (its a bit verbose now)?

```
const uint64_t current_time = offset_time.valid ? offset_time.local_time : 0;
const uint32_t tz_index = offset_time.valid ? nodeinfo->timezone() : 0;

const bool allowed =
          FORWARD ? costing_->Allowed(directededge, pred, tile, edgeid, current_time,
                                      tz_index, restriction_idx)
                  : costing_->AllowedReverse(directededge, pred, opp_edge, t2, oppedgeid,
                                             current_time, tz_index, restriction_idx);
if (!allowed || costing_->Restricted(directededge, pred, bdedgelabels_, tile, edgeid, true,
                                           todo, current_time, tz_index) {
...
```

Reply:
The end goal is to merge this with *other* implementations of this same logic - I'd like to keep the code *mostly* the same for now to make it easier to diff against the other code.